core_tests: test reorg with alternative deregisters

msgmaxim opened the pull request.

Doy-lee commented on tests/core_tests/service_nodes.h:
```diff
@@ -134,4 +134,41 @@ struct get_test_options<test_deregister_safety_buffer> {
+{
+  const std::pair<uint8_t, uint64_t> hard_forks[3] = { std::make_pair(7, 0),
+                                                       std::make_pair(8, 1),
+                                                       std::make_pair(9, 2) };
```
This causes an array overrun here:
https://github.com/loki-project/loki/blob/ffdf6826f77398520e69736032f2564d65afa9d5/src/cryptonote_core/blockchain.cpp#L336

Similarly for the other tests where you declare the hard-forks. The code in cryptonote_core iterates over hard_forks expecting that it is null terminated. So what hardforks should look like is 

```
const std::pair<uint8_t, uint64_t> hard_forks[4] = 
{ 
    std::make_pair(7, 0),
    std::make_pair(8, 1),
    std::make_pair(9, 2),
    std::make_pair(0, 0),
};
```

I think there's a good case in making cryptonote::test_options have int num_hard_forks, instead of the implicit knowledge, this has to be null terminated.

Doy-lee replied:
Do we have an update on this?

msgmaxim commented:
Rebased and changed from a pointer to array of options to a vector to keep track of the size.

I also added the `--service_nodes` option for core tests that will only run service node tests. By default, it will run all other tests.
